Source/MQTTnet/MqttApplicationMessageBuilder.cs

```diff
@@ -112,76 +119,89 @@ public MqttApplicationMessageBuilder WithMessageExpiryInterval(uint messageExpir
         public MqttApplicationMessageBuilder WithPayload(byte[] payload)
         {
             _payload = payload;
+            _payloadOffset = 0;
+            _payloadCount = null;
+            return this;
+        }
+
+        public MqttApplicationMessageBuilder WithPayload(ArraySegment<byte> payload)
+        {
+            _payload = payload.Array;
+            _payloadOffset = payload.Offset;
+            _payloadCount = payload.Count;
             return this;
         }
 
         public MqttApplicationMessageBuilder WithPayload(IEnumerable<byte> payload)
         {
             if (payload == null)
             {
-                _payload = null;
-                return this;
+                return WithPayload(default(byte[]));
             }
 
-            _payload = payload as byte[] ?? payload.ToArray();
+            if (payload is byte[] byteArray)
+            {
+                return WithPayload(byteArray);
+            }
 
-            return this;
+            if (payload is ArraySegment<byte> arraySegment)
+            {
+                return WithPayload(arraySegment);
+            }
+
+            return WithPayload(payload.ToArray());
         }
 
         public MqttApplicationMessageBuilder WithPayload(Stream payload)
         {
-            if (payload == null)
-            {
-                _payload = null;
-                return this;
-            }
-
-            return WithPayload(payload, payload.Length - payload.Position);
+            return payload == null
+                ? WithPayload(default(byte[]))
+                : WithPayload(payload, payload.Length - payload.Position);
         }
 
         public MqttApplicationMessageBuilder WithPayload(Stream payload, long length)
         {
-            if (payload == null)
+            if (payload == null || length == 0)
             {
-                _payload = null;
-                return this;
+                return WithPayload(default(byte[]));
             }
 
-            if (payload.Length == 0)
-            {
-                _payload = null;
-            }
-            else
+            var payloadBuffer = new byte[length];
+            var totalRead = 0;
+            do
```

Comment:
FYI: for .NET 7 onwards `ReadExactly` can be used. But ATM there's no such target https://github.com/dotnet/MQTTnet/blob/7f4a7ef668c15ba968a9950cca4898bf47a776c1/Source/MQTTnet/MQTTnet.csproj#L13

Reply:
Yes, but the original logic is retained here, just modified to be compatible with the new PayloadOffset and PayloadLength propties.